File: examples/cpp/rotation_sweep_custom/rotation_sweep_custom_example.cpp
#include <qristal/core/benchmark/workflows/RotationSweep.hpp>
#include <qristal/core/benchmark/workflows/PreOrAppendWorkflow.hpp>
#include <qristal/core/benchmark/metrics/BitstringCounts.hpp>
#include <qristal/core/primitives.hpp>

#include <time.h>
#include <range/v3/view/zip.hpp>

using namespace qristal::benchmark;

int main() {

    const size_t n_qubits = 1;

    //(1) set up session
    qristal::session sim;
    sim.acc = "qpp";
    sim.sn = 1000;
    sim.qn = n_qubits;

    //(2) define workflow 
    RotationSweep base_workflow({'Z'}, -180, +180, 5, sim);
    //Prepend rotation to |X+> to workflow circuits 
    std::vector<qristal::BlochSphereUnitState> prepend {qristal::BlochSphereUnitState::Symbol::Xp};
    PreOrAppendWorkflow<RotationSweep> prepended_workflow(
        base_workflow, 
        prepend, 
        Placement::Prepend
    );
    //Append measurements (X, Z) to workflow circuits
    std::vector<std::vector<qristal::Pauli>> append {
        {qristal::Pauli::Symbol::X}, 
        {qristal::Pauli::Symbol::Z}
    };
    PreOrAppendWorkflow<PreOrAppendWorkflow<RotationSweep>> final_workflow(
        prepended_workflow, 
        append, 
        Placement::Append
    );

    //(3) pass to metric
    BitstringCounts<PreOrAppendWorkflow<PreOrAppendWorkflow<RotationSweep>>> metric(final_workflow);
    std::vector<qristal::CircuitBuilder> circuits = final_workflow.get_circuits();

    //(4) evaluate and print
    auto results = metric.evaluate();
    for (const auto& [timestamp, counts_vec] : results) {
        std::cout << "Evaluated metric from " << "UTC: " << std::put_time(std::gmtime(&timestamp), "%c %Z")
                  << "(local: " << std::put_time(std::localtime(&timestamp), "%c %Z") << "): [" << std::endl;
        for (const auto& [counts, circuit] : ::ranges::views::zip(counts_vec, circuits)) {
            std::cout << "Circuit:" << std::endl; 
            circuit.print(); 
            std::cout << "Measured bitstring counts:" << std::endl; 
            std::cout << counts << std::endl;
            std::cout << "---" << std::endl;
        }
    }

}

File: examples/python/benchmark/rotation_sweep_custom_example.py
import qristal.core
import qristal.core.benchmark as benchmark

from datetime import datetime, timezone
import numpy as np

n_qubits = 1
n_shots = 1000

#(1) define session
sim = qristal.core.session()
sim.acc = "qpp"
sim.sn = n_shots
sim.qn = n_qubits

#(2) define workflow
base_workflow = benchmark.RotationSweep(['Z'], -180, +180, 5, sim)
#Prepend rotation to |X+> to workflow circuits 
prepend = [benchmark.BlochSphereUnitState(benchmark.BlochSphereUnitStateSymbol.Xp)]
prepended_workflow = benchmark.PreOrAppendWorkflow(base_workflow, prepend, benchmark.Placement.Prepend)
#Append measurements (X, Z) to workflow circuits
append = [
    [benchmark.Pauli(benchmark.PauliSymbol.X)], 
    [benchmark.Pauli(benchmark.PauliSymbol.Z)]
]
final_workflow = benchmark.PreOrAppendWorkflow(prepended_workflow, append, benchmark.Placement.Append)

#(3) pass to metric
metric = benchmark.BitstringCounts(final_workflow) 

#(4) evaluate and print
results = metric.evaluate() 
for timestamp, counts_vec in results.items(): 
    # Convert timestamp to UTC and local time
    utc_time = datetime.fromtimestamp(timestamp, tz=timezone.utc)
    local_time = datetime.fromtimestamp(timestamp)  # Local timezone
    print(f"Evaluated metric from UTC: {utc_time.strftime('%c %Z')} "
          f"(local: {local_time.strftime('%c %Z')}):")
    for counts, circuit in zip(counts_vec, final_workflow.circuits):
        print("Circuit:") 
        circuit.print() 
        print("Measured bitstring counts:")
        print(counts)
        print("---")
